File: app/schemas/recording.py
from pydantic import BaseModel, Field
from typing import Optional, List, Union, Dict, Any
from enum import Enum

class CleanupPolicyType(str, Enum):
    COUNT = 'count'
    SIZE = 'size'
    AGE = 'age'
    CUSTOM = 'custom'

class PreserveTimeframeSchema(BaseModel):
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    weekdays: Optional[List[int]] = None
    timeOfDay: Optional[Dict[str, str]] = None

class CleanupPolicySchema(BaseModel):
    type: CleanupPolicyType = CleanupPolicyType.COUNT
    threshold: int = Field(default=10, description="Threshold value for the policy")
    preserve_favorites: bool = Field(default=True, description="Preserve recordings of favorite categories")
    preserve_categories: Optional[List[str]] = Field(default=None, description="Categories to preserve")
    preserve_timeframe: Optional[PreserveTimeframeSchema] = Field(default=None, description="Timeframe to preserve")

class RecordingSettingsSchema(BaseModel):
    enabled: bool = Field(default=True, description="Enable recording globally")
    output_directory: str = Field(default="/recordings", description="Directory to save recordings")
    filename_template: str = Field(default="{streamer}/{streamer}_{year}-{month}-{day}_{hour}-{minute}_{title}_{game}", 
                                  description="Template for recording filenames")
    filename_preset: Optional[str] = Field(default="default", description="Preset template for recording filenames")
    default_quality: str = Field(default="best", description="Default recording quality")
    use_chapters: bool = Field(default=True, description="Create chapters based on stream events")
    use_category_as_chapter_title: bool = Field(default=False, description="Use category name as chapter title instead of stream title")
    max_streams_per_streamer: int = Field(default=0, description="Maximum number of streams to keep per streamer (0 = unlimited)")
    cleanup_policy: Optional[CleanupPolicySchema] = None

    class Config:
        from_attributes = True

class StreamerRecordingSettingsSchema(BaseModel):
    streamer_id: int
    username: Optional[str] = None
    profile_image_url: Optional[str] = None
    enabled: bool = True
    quality: Optional[str] = None
    custom_filename: Optional[str] = None
    max_streams: Optional[int] = None
    cleanup_policy: Optional[CleanupPolicySchema] = None

    class Config:
        from_attributes = True
        
class ActiveRecordingSchema(BaseModel):
    streamer_id: int
    streamer_name: str
    started_at: str
    duration: float
    output_path: str
    quality: str
    
class StorageUsageSchema(BaseModel):
    totalSize: int
    recordingCount: int
    oldestRecording: str
    newestRecording: str
    
class CleanupResultSchema(BaseModel):
    status: str
    message: str
    deleted_count: int
    deleted_paths: List[str]

File: app/frontend/src/types/recording.ts
export interface RecordingSettings {
  enabled: boolean;
  filename_template: string;
  filename_preset?: string;
  default_quality: string;
  use_chapters: boolean;
  use_category_as_chapter_title?: boolean; // Neue Eigenschaft hinzufügen
  cleanup_policy?: CleanupPolicy; // New property for cleanup policy
}

export interface StreamerRecordingSettings {
  streamer_id: number;
  username: string;
  enabled: boolean;
  quality?: string;
  custom_filename?: string;
  profile_image_url?: string;
  max_streams?: number; // Maximum number of streams to keep for this streamer
  cleanup_policy?: CleanupPolicy; // New property for streamer-specific cleanup policy
  use_global_cleanup_policy?: boolean; // Use global cleanup policy or streamer-specific (default: true)
}

// New type for advanced cleanup policies
export interface CleanupPolicy {
  type: CleanupPolicyType;
  threshold: number;
  preserve_favorites: boolean;
  preserve_categories: string[];
  preserve_timeframe?: PreserveTimeframe; // Made optional to simplify
}

export enum CleanupPolicyType {
  COUNT = 'count',       // Limit by number of recordings
  SIZE = 'size',         // Limit by total size (in GB)
  AGE = 'age',           // Limit by age (in days)
  CUSTOM = 'custom'      // Custom policy (combination of the above)
}

export interface PreserveTimeframe {
  start_date: string;   // ISO date string
  end_date: string;     // ISO date string 
  weekdays: number[];   // 0-6 (Sunday-Saturday)
  timeOfDay: {
    start: string;       // HH:MM format
    end: string;         // HH:MM format
  };
}

export interface ActiveRecording {
  id: number;              // recording_id
  stream_id: number;
  streamer_id: number;
  streamer_name: string;
  title: string;
  started_at: string;
  file_path: string;
  status: string;
  duration: number;
  // Legacy fields for backwards compatibility
  output_path?: string;
  quality?: string;
}

export const QUALITY_OPTIONS = [
  { value: "best", label: "Best Available" },
  { value: "1080p60", label: "1080p (60 fps)" },
  { value: "1080p", label: "1080p (30 fps)" },
  { value: "720p60", label: "720p (60 fps)" },
  { value: "720p", label: "720p (30 fps)" },
  { value: "480p", label: "480p" },
  { value: "360p", label: "360p" },
  { value: "160p", label: "160p" },
  { value: "audio_only", label: "Audio Only" }
];

export const FILENAME_VARIABLES = [
  { key: "streamer", description: "Streamer username" },
  { key: "title", description: "Stream title" },
  { key: "game", description: "Game/category name" },
  { key: "twitch_id", description: "Twitch ID" },
  { key: "year", description: "Year (YYYY)" },
  { key: "month", description: "Month (MM)" },
  { key: "day", description: "Day (DD)" },
  { key: "hour", description: "Hour (HH)" },
  { key: "minute", description: "Minute (MM)" },
  { key: "second", description: "Second (SS)" },
  { key: "timestamp", description: "Timestamp (YYYYMMDD_HHMMSS)" },
  { key: "datetime", description: "Formatted datetime (YYYY-MM-DD_HH-MM-SS)" },
  { key: "id", description: "Stream ID" },
  { key: "season", description: "Season format (S2024-01)" },
  { key: "episode", description: "Episode number (E01, E02, ...)" },
  { key: "unique", description: "Unique identifier" }
];

export interface FilenamePreset {
  value: string;
  label: string;
  description: string;
}

export const FILENAME_PRESETS: FilenamePreset[] = [
  { 
    value: "default", 
    label: "Default", 
    description: "{streamer}/{streamer}_{year}-{month}-{day}_{hour}-{minute}_{title}_{game}"
  },
  { 
    value: "plex", 
    label: "Plex/Emby/Jellyfin", 
    description: "{streamer}/Season {year}-{month}/{streamer} - S{year}{month}E{episode:02d} - {title}"
  },
  { 
    value: "emby", 
    label: "Emby Alternative", 
    description: "{streamer}/Season {year}-{month}/{streamer} - S{year}{month}E{episode:02d} - {title}"
  },
  { 
    value: "jellyfin", 
    label: "Jellyfin Alternative", 
    description: "{streamer}/Season {year}-{month}/{streamer} - S{year}{month}E{episode:02d} - {title}"
  },
  { 
    value: "kodi", 
    label: "Kodi", 
    description: "{streamer}/Season {year}-{month}/{streamer} - S{year}{month}E{episode:02d} - {title}"
  },
  { 
    value: "chronological", 
    label: "Chronological", 
    description: "{year}/{month}/{day}/{streamer} - E{episode:02d} - {title} - {hour}-{minute}"
  }
];
